spec/GraphQLOverHTTP.md

```diff
@@ -201,8 +201,38 @@ may not be the GraphQL server but instead some intermediary such as API gateways
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
-It is recommended to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+If the response has Content-Type GraphQL and contains a non-null `data` entry,
+then it MUST have status code `2xx`, and it SHOULD have status code `200`.
+
+If the response has Content-Type GraphQL and has status code `2xx`,
+the entry `data` entry must be either:
+- equal to `null` (in case of an execution error)
+- not present (in case of a validation error)
+
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation typically become part of the result,
+as long as the server is still able to produce a well-formed response.
+
+In case of errors that completely prevent the successful execution of the request,
+the server SHOULD respond with the appropriate status code depending on the concrete error condition.
+
+There are four types of validation errors:
+
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).
+
+The error types 1. and 2. prevent execution of the request entirely and MUST result in
+status code `400` (Bad Request). Types 3. and 4. may still allow partial execution and
+MUST result in status code `200`.
```

Comment:
This currently disagrees with the flexibility of the wording above. I've softened it:

```suggestion
Error types 1. and 2. prevent execution of the request in its entirety; they MUST result in
status code `4xx` or `5xx`, and SHOULD result in status code `400` (Bad Request). Types 3. and 4. may still allow partial execution; they
MUST result in status code `2xx`, and SHOULD result in status code `200` (Okay).
```

Reply:
Mutations are usually atomic operations so the case of partial execution doesn't exist (except for say chained mutations).  How about leaving it to the application's business logic to determine the best HTTP error code suitable for the use case, without being prescriptive about which exact error code should it throw. For example, incase of idempotent mutation, for conflict scenarios you throw 409. downstream connection failure errors while executing a mutation can throw 500 and the upstream client can retry the request using exponential backoff.

Reply:
The selection set on a mutation field is effectively a query, so partial execution applies to mutation operations in just the same way it does to query operations.

Further, mutation operations can contain more than one mutation (field) within them, so using a code such as 409 does not make sense - which of the mutations does it apply to? If one succeeded and one failed, do we use the 409 for the failure or the 200 for the success?

```graphql
# Assuming this results in a 409 status; which one of these went wrong? Did they both fail?
mutation Conflict {
  createUser(username: "Bob") { id name }
  updateUser(username: "Alice", toUsername: "Alan") { id name }
}
```

We should do our best to help people avoid these ambiguities, that's our job writing the spec - to make sure that all situations are handled and clear. This is why we've stated that as soon as execution is successfully invoked, the response should be a 200 status (successful GraphQL execution) and it is then up to the client to process the results of the execution which it is best placed to do, given it understands GraphQL and knows what request it sent.

Reply:
I think we can separate a purely atomic mutation operation (that can either fail or pass- almost all mutations fall in this category) from a composite mutation operation (which involves multiple steps) or a chained mutation like the example you gave above (composite or chained mutations are used only in special cases).  The idempotent mutation example that I gave above is an atomic mutation (for example `makePayment`) and considering that `makePayment` is declared idempotent, when `makePayment` is already processing and another duplicate request arrives, the error that  is appropriate to throw is a 409 Conflict. Throwing a 200 here doesn't help. Same with 500 errors when there's a downstream connection failure/timeout, while processing the request during the graphql execution). The intermediary proxies, monitoring systems in any enterprise infrastructure won't introspect the response body to log the errors, create appropriate alerts or do automatic retries in case of 500 errors and so on. Throwing an HTTP error code as-is from atomic mutation operation will make the existing web infra work as-is. Neither the API provider nor the API consumer need to write any custom logic for understanding GraphQL responses. This is the single most complaint (GraphQL error handling) I hear from most clients on why they don't want to consume or build **public** GraphQL APIs. I think we have an opportunity to address that gap through this working group. Solving this problem will help towards more providers building GraphQL public APIs.

My recommendation (If these make sense) :

1- We allow the API provider to emit one of the standard HTTP status codes as appropriate without being prescriptive about which code to use when or instead of saying use 200 only if the error occurs during graphql execution. Instead we let the application decide on which status code to emit (e.g use 200 for chained mutations if there are partial errors or appropriate 4XX/5XX when there= are errors, while executing atomic mutations), regardless of whether the error occurred prior to execution or during execution.

2- We should think this specification as the HTTP binding spec for GraphQL, which means, we should let the HTTP protocol dictates the rules. That said we should keep the core GraphQL constructs as-is (of course we shouldn't allow someone to use a PUT method for a mutation:-)) . As an example from another spec we can check the [CNCF Cloudevent Specification](https://github.com/cloudevents/spec). The Cloudevent envelope format is a protocol agnostic envelope to describe events and CNCF defines binding rules for each type of transport (Including HTTP). See all binding specs [here](https://github.com/cloudevents/spec)  and let the protocol dictate the status code, security etc.

3- We should define some http specific directives to define HTTP metadata (like request/response headers, declaring a method is idempotent, status codes, produces/consumes etc.

Reply:
To summarize, I think you're saying:

If:

- the operation is a mutation, and
- the operation contains exactly one non-introspection (i.e. not `__typename`) root-level field, and
- that field throws an error

Then:

- the server MAY choose to throw a non-200 status code

Else:

- the server MUST result in status code `2xx`, and SHOULD result in status code `200` (Okay)`

Your example of 500 errors getting auto-retried on the back-end is where I've drawn these rules from; for example if the field `name` in my `Conflict` example above were to fail due to network connectivity (but all other fields, including the mutation itself, succeeded) then presumably this should not auto-retry since it's an expected failure state that GraphQL was designed to handle gracefully with its nullable-by-default partial-response + errors.

If this is an inaccurate description of your position, please could you expand exactly when you think non-200 status codes should be valid?

---

Separately, how do you intend to indicate that a specific mutation field is idempotent, and which part of the stack is the consumer of that information?

Reply:
@benjie almost. Let me add some examples for atomic mutations and illustrate the scenarios where the client always executes one mutation at a time

**Make Payment**:

~~~
 {
     "query" : "mutation makePayment (...)"
     "variables" : {...}

}
~~~

**Create User**:

~~~
 {
     "query" : "mutation createUser (...)"
     "variables" : {...}

}
~~~

In the examples above;

* Both the Mutations `makePayment` and `createUser` are atomic and do only one atomic action.
* Both of the above mutations are always executed independently but never together
* There's no chance of partial failure here. Any of the mutations when executed either fail or succeed.


Considering the above, here are the error scenarios;
* `makePayment` can return 400, when the receiver's emai ld/phone number is incorrect or account number is invalid or aaccount is inactive and payment cant be made.

* `makePayment` is idempotent (implements exactly once semantics) so when you send a duplicate `makePayment` call with same payload again, the server would reject the call with a 409 Conflict error.

* When `makePayment` cant reach the card network server (Visa or Master card) that the sender used the card to make payment or the request with card network times out, `makePayment` would like to return a 500/504 error, as its an intermittent recoverable error and the likelihood of payment being successful if it is retried again is very high.Returning a 500 to the API Gateway or API client, will make them retry the request. That is how all intermediary proxies are configured to do automatic retries on 500s using exponential backoff.

* `makePayment` may return a 403 Authorization error if the user is not allowed to make payments (say its a child account and parents have restricted the child account permission to only receive payments but not make any payments)

*A monitoring system will send alerts to customer support team if there are too many payment failures. Monitoring systems look for HTTP error codes. Similarly a monitoring system will report to security incident team if tehre are too many authorization errors (say 403). 

* A rate limiter system may block requests if there are too many requests from a device that fail with authorization errors (Rate limiter systems depend on HTTP status codes).

So all atomic operations should be able to return 4XX/5XX errors as appropriate and that makes the errors explicit and clear and uses HTTP protocol fully. Any traditional client with standard HTTP infrastructure doesn't need to write any custom error handling code to integrate with GraphQL.

**The other Use cases (which result in Partial failures so 200 response is OK):**
In the case of chained mutations (your example above, there are always chances of partial errors so a 200 response with `data` and `errors` are appropriate.

To conclude, my recommendation is that we allow the GraphQL service implementations return appropriate HTTP status codes, without being prescriptive about when to throw 4XX/5XX vs when to throw 2XX. 

Rewarding idempotency, you can for example, come up with a directive such as `@idempotent`. When you declare a method idempotent it helps client to retry such method automatically in case of request timeouts. on the server side, its natural that the implementation enforces a constraint wherein all clients need to send an unique idempotency key for every mutation call that is declared idempotent (this is a standard industry pattern). If the client doesnt send a idempotency-key in the mutation call, the server rejects the request with a 400 error.

Example:

~~~
type mutation {

    makePayment(...) @idempotent

}
~~~

Hope this helps.

Reply:
> Both of the above mutations are always executed independently but never together

This seems to be a human rule rather than a computer rule. How can the computer know this? How would it know the following operation is forbidden?

```graphql
mutation TakePaymentAndCreateAccount ($accountDetails: CreateAccountInput!, $paymentDetails: TakePaymentInput!) {
  takePayment(input: $paymentDetails) { success errors { code message } }
  createAccount(input: $accountDetails) { userId }
}
```

It's very common during a checkout process for the website to also invite you to create an account. How can we expand this to GraphQL in general that's not specific to your server's setup?

---

Regarding your example of the `@idempotent` directive on a mutation, there's currently no way for baseline Spec compliant GraphQL servers to indicate that a mutation has this tag; it's something that I hope we will have in the coming years but work on that is ongoing. So the server cannot indicate this; and further even if the server _did_ indicate this _through GraphQL_ then for an intermediary service to understand it, it would need to understand GraphQL, at which point using the HTTP status code for auto-retry becomes unnecessary since the gateway can extract the information it needs from the GraphQL request directly (although slight rewriting of the query might be necessary if the user did not request sufficient information). The same "proxy server needs to understand GraphQL" applies also to the case of a client-indicated `@idempotent` directive which is what your example demonstrates (although I'm not sure if it is what you meant?).

Reply:
(I note "baseline Spec compliant GraphQL servers" because some implementations (GraphQL Java?) have experimental support for indicating directives applied to fields within the introspection results.)

Reply:
> There's no chance of partial failure here. Any of the mutations when executed either fail or succeed.

The fact that mutations also return a selection set makes this problematic. Let's expand upon the example query:

```graphql
mutation {
  createAccount(input: $accountDetails) {
    id
    friends {
      someAccountDetailOnlyAvailableToAdmins
    }
  }
}
```

Suppose you were to return an error code 403; how can the client differentiate between:
- they are not allowed to create accounts
- they are not an admin and cannot view certain fields

A single status code provides too little information here. Given the complex nature of queries, reducing the result down to a single status code simply is not feasible once execution started. The HTTP semantics for looking at the status codes of individual requests simply falls apart, since it is based upon the presumption that *one request does one thing and one thing only*.

GraphQL does provide means of communicating errors within the execution within the result. That is by design. I get that adopting such a paradigm shift may be challenging and hinder adoption, but shoehorning HTTP semantics on top will not help in the long run.

Reply:
@benjie thanks for the suggestion, I added it in but will leave this discussion open for now.